### DISCUSSION TITLE
How to deserialize in struct with null values?

### DISCUSSION THREAD

**Discussion by Heavenfighter** [nlohmann/json#3185] [Q&A] [answered]
I'm getting following json from an webservice:

```cpp
json j2 = R"(
	{
		"Item": {
			"Attributes": [
				{
				"Name": "Text00",
				"DisplayName": "ClientCalling",
				"Value": "Incomming"
				}
			],
			"ObjectType": {
			  "Id": 76,
			  "Name": null
			}
		}
	})"_json;
```

My structs looks like:

```cpp
struct Attribute
{
std::string Name;
std::string DisplayName;
std::string Value;
};
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Attribute, Name, DisplayName, Value)

struct Item
{
int Id;
std::string Name;
};
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Item, Id, Name)

struct ObjectTypeTestItem
{
std::vector <Attribute> Attributes;
Item ObjectType;
};
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(ObjectTypeTestItem, Attributes, ObjectType)
```

I've tried to deserialize with

```cpp
auto t = j2["Item"].get<WDObjectTypeTestItem>();
```

but i'm getting following exception:
` Microsoft C++ exception: nlohmann::detail::type_error at memory location 0x0A7CE278.`
	
It seems that the cause is the null value in name attribute of the ObjectType member.

What can i do to deserialize this json correctly?

**Comment by gregmarr** [comment, marked as answer]
You need to write the from_json function yourself in this case.  There is a PR a version of these macros that handles missing elements, but it doesn't handle "wrong type" elements such as being null rather than string or not provided.

**Reply by Heavenfighter** [reply]
thank you, thats what works
